lib/Exporter/Writer/KnpSnappyWriter.php

```diff
@@ -0,0 +1,221 @@
+    <style></style>
+  </head>
+  <body>
+    [header-html]
+    <table>[content]</table>
+    [footer-html]
+  </body>
+</html>';
+
+    /**
+     * @throws \RuntimeException
+     *
+     * @param string $filename
+     * @param bool   $showHeaders
+     * @param bool   $showBorders
+     * @param array  $snappyOptions     KnpSnappy options (wkhtmltopdf)
+     * @param string $wkhtmltopdfBinary Path of the wkhtmltopdf binary
+     */
+    public function __construct(
+        $filename,
+        $wkhtmltopdfBinary,
+        $showHeaders = true,
+        $showBorders = true,
+        array $snappyOptions = array()
+    ) {
+        $this->filename = $filename;
+        $this->showHeaders = $showHeaders;
+        $this->showBorders = $showBorders;
+        $this->snappyOptions = $snappyOptions;
+        $this->wkhtmltopdfBinary = $wkhtmltopdfBinary;
+
+        if (is_file($filename)) {
+            throw new \RuntimeException(sprintf('The file "%s" already exists', $filename));
+        }
+
+        if (!is_executable($wkhtmltopdfBinary)) {
+            throw new \RuntimeException('The file "%s" doesn\'t exist or is not executable', $wkhtmltopdfBinary);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function open()
+    {
+        $this->file = fopen($this->filename, 'w', false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function write(array $data)
+    {
+        $tmpFile = tmpfile();
+
+        if (true === $this->showBorders) {
+            $this->addBorders();
+        }
+
+        $this->addMarkupFromOptions();
+
+        if (true === $this->showHeaders) {
+            fwrite($tmpFile, $this->getHeaders($data));
+        }
+
+        fwrite($tmpFile, '<tr>');
+
+        foreach ($data as $value) {
+            fwrite($tmpFile, sprintf('<td>%s</td>', $value));
+        }
+
+        fwrite($tmpFile, '</tr>');
+
+        $tmpFileMetaData = stream_get_meta_data($tmpFile);
+        $tmpFilePath = $tmpFileMetaData['uri'];
+
+        fwrite($this->file, $this->generatePdf($tmpFilePath));
+
+        fclose($tmpFile);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function close()
+    {
+        fclose($this->file);
+    }
+
+    /**
+     * Generates a PDF using KnpSnappy.
+     *
+     * @param string $htmlFilename.
+     *
+     * @return string The generated PDF content.
+     */
+    private function generatePdf($htmlFilename)
+    {
+        $snappy = new Pdf();
+        $snappy->setBinary($this->wkhtmltopdfBinary);
+
+        foreach ($this->snappyOptions as $key => $value) {
+            $snappy->setOption($key, $value);
+        }
+
+        $this->html = str_replace('[content]', file_get_contents($htmlFilename), $this->html);
+
+        return $snappy->getOutputFromHtml($this->html);
+    }
+
+    /**
+     * @param array $data
+     *
+     * @return string
+     */
+    private function getHeaders(array $data)
```

Comment:
`$data` is used by `WriterInterface::write()` and all the Writer classes (even for methods such as "addHeaders". I really don't see how I can be more precise without being wrong.